signature/jws/signer.go

```diff
@@ -68,31 +55,99 @@ func NewSignerWithCertificateChain(method jwt.SigningMethod, key crypto.PrivateK
+	if err != nil {
+		return nil, err
+	}
+
 	rawCerts := make([][]byte, len(certChain))
 	for i, cert := range certChain {
 		rawCerts[i] = cert.Raw
 	}
 	return &Signer{
-		method:    method,
-		key:       key,
-		certChain: rawCerts,
+		runner: &builtinPlugin{
+			keySpec:   keySpec,
+			method:    method,
+			key:       key,
+			certChain: rawCerts,
+		},
 	}, nil
 }
 
-// Sign signs the artifact described by its descriptor, and returns the signature.
-func (s *Signer) Sign(ctx context.Context, desc notation.Descriptor, opts notation.SignOptions) ([]byte, error) {
-	// generate JWT
-	payload := packPayload(desc, opts)
-	if err := payload.Valid(); err != nil {
-		return nil, err
+// builtinPlugin is a plugin.Runner implementation which
+// signs supports the generate-signature workflow using
+// the provided key and certificates.
+type builtinPlugin struct {
+	keySpec notation.KeySpec
+
+	// method is the method to sign artifacts.
+	method jwt.SigningMethod
+
+	// key is the signing key used to sign artifacts.
+	// The signing key can be either remote or local.
+	key crypto.PrivateKey
+
+	// certChain contains the X.509 public key certificate or certificate chain corresponding
+	// to the key used to generate the signature.
+	certChain [][]byte
+}
+
+func (builtinPlugin) metadata() *plugin.Metadata {
+	// The only properties that are really relevant
+	// are the supported contract version and the capabilities.
+	// All other are just filled with meaningful data.
+	return &plugin.Metadata{
+		SupportedContractVersions: []string{plugin.ContractVersion},
+		Capabilities:              []plugin.Capability{plugin.CapabilitySignatureGenerator},
+		Name:                      "built-in",
+		Description:               "Built-in signer",
+		Version:                   plugin.ContractVersion,
+		URL:                       "https://github.com/notaryproject/notation-go",
 	}
-	token := jwtToken(s.method.Alg(), payload)
-	token.Method = s.method
-	compact, err := token.SignedString(s.key)
-	if err != nil {
-		return nil, err
+}
+
+// Run implement the generate-signature workflow.
+func (r *builtinPlugin) Run(ctx context.Context, req plugin.Request) (interface{}, error) {
+	switch req.Command() {
+	case plugin.CommandGetMetadata:
+		return r.metadata(), nil
+	case plugin.CommandDescribeKey:
+		req1 := req.(*plugin.DescribeKeyRequest)
+		return &plugin.DescribeKeyResponse{
+			KeyID:   req1.KeyID,
+			KeySpec: r.keySpec,
+		}, nil
+	case plugin.CommandGenerateSignature:
+		req1 := req.(*plugin.GenerateSignatureRequest)
+		// TODO: the builtinPlugin should be JWS-agnostic.
+		// Stop using a jwt.MethodSigner and use instead
+		// the hash provided in req1.Hash and a Sign method
+		// which does not hash data itself.
+		signed, err := r.method.Sign(string(req1.Payload), r.key)
+		if err != nil {
+			return nil, plugin.RequestError{
+				Code: plugin.ErrorCodeGeneric,
+				Err:  err,
+			}
+		}
+		signedDecoded, err := base64.RawURLEncoding.DecodeString(signed)
```

Comment:
I'd add a comment about why we need to decode the signature before sending it in response.

Reply:
Done